flang/lib/Optimizer/Transforms/OMPMapInfoFinalization.cpp

```diff
@@ -85,67 +138,187 @@ class OMPMapInfoFinalizationPass
+    // appropriate position to maintain correctness of the positional/index data
+    // for that member.
+    size_t insertPosition =
+        std::distance(baseAddrIndex.begin(), std::prev(baseAddrIndex.end()));
+    for (size_t i = 0; i < memberIndices.size(); ++i) {
+      if (memberIndices[i].size() > insertPosition &&
+          std::equal(baseAddrIndex.begin(), std::prev(baseAddrIndex.end()),
+                     memberIndices[i].begin())) {
+        memberIndices[i].insert(
+            std::next(memberIndices[i].begin(), insertPosition), 0);
+      }
+    }
 
-      for (size_t i = 0; i < mapVarsArr.size(); ++i) {
-        if (mapVarsArr[i] == op) {
-          // Push new implicit maps generated for the descriptor.
-          newMapOps.push_back(baseAddr);
+    // Insert our newly created baseAddrIndex into the larger list of indices at
+    // the correct location.
+    memberIndices.insert(std::next(memberIndices.begin(), memberIndex + 1),
+                         baseAddrIndex);
+  }
 
-          // for TargetOp's which have IsolatedFromAbove we must align the
-          // new additional map operand with an appropriate BlockArgument,
-          // as the printing and later processing currently requires a 1:1
-          // mapping of BlockArgs to MapInfoOp's at the same placement in
-          // each array (BlockArgs and MapVars).
-          if (auto targetOp = llvm::dyn_cast<mlir::omp::TargetOp>(target))
-            targetOp.getRegion().insertArgument(i, baseAddr.getType(), loc);
-        }
-        newMapOps.push_back(mapVarsArr[i]);
+  /// Adjusts the descriptors map type the main alteration that is done
+  /// currently is transforming the map type to OMP_MAP_TO where possible.
+  // This is because we will always need to map the descriptor to device
+  /// (or at the very least it seems to be the case currently with the
+  /// current lowered kernel IR), as without the appropriate descriptor
+  /// information on the device there is a risk of the kernel IR
+  /// requesting for various data that will not have been copied to
+  /// perform things like indexing, this can cause segfaults and
+  /// memory access errors. However, we do not need this data mapped
+  /// back to the host from the device, as we cannot alter the data
+  /// via resizing or deletion on the device, this is specified in the
+  /// OpenMP specification, so discarding any descriptor alterations via
+  /// no map back is reasonable (and required for certain segments
+  /// of descriptor data like the type descriptor that are global
+  /// constants). This alteration is only unapplicable to
+  /// target exit and target update currently, and that's due to
+  /// target exit not allowing To mappings, and target update not
+  /// allowing both to and from simultaneously. We currently try
+  /// to maintain the implicit flag where neccesary, although, it
+  /// does not seem strictly required.
```

Comment:
Nit: These are mainly some suggestions to punctuation, since I think it's a bit difficult to follow at the moment. Also re-flow comments, since they seem to not be extending to 80 characters.
```suggestion
  /// Adjusts the descriptor's map type. The main alteration that is done
  /// currently is transforming the map type to `OMP_MAP_TO` where possible.
  /// This is because we will always need to map the descriptor to device
  /// (or at the very least it seems to be the case currently with the
  /// current lowered kernel IR), as without the appropriate descriptor
  /// information on the device there is a risk of the kernel IR
  /// requesting for various data that will not have been copied to
  /// perform things like indexing. This can cause segfaults and
  /// memory access errors. However, we do not need this data mapped
  /// back to the host from the device, as per the OpenMP spec we cannot alter the data
  /// via resizing or deletion on the device. Discarding any descriptor alterations via
  /// no map back is reasonable (and required for certain segments
  /// of descriptor data like the type descriptor that are global
  /// constants). This alteration is only inapplicable to
  /// `target exit` and `target update` currently, and that's due to
  /// `target exit` not allowing `to` mappings, and `target update` not
  /// allowing both `to` and `from` simultaneously. We currently try
  /// to maintain the `implicit` flag where necessary, although it
  /// does not seem strictly required.
```